IllegalStateException when creating bundle packet container in 1.19.4

- [x] This issue is not solved in a development build

**Describe the bug**
When writing to the packet bundles structure of a PacketContainer of the type `PacketType.Play.Server.BUNDLE` in a 1.19.4 server, ProtocolLib throws an exception.

The issue appears to be related to a failed cast from `net.minecraft.network.protocol.BundleDelimiterPacket` to `net.minecraft.network.protocol.BundlePacket`.

**To Reproduce**
```java
List<PacketContainer> packets = new ArrayList<>();
for (int i=0; i<10; i++) {
    PacketContainer packet = new PacketContainer(PacketType.Play.Server.SET_ACTION_BAR_TEXT);
    packet.getChatComponents().write(0, WrappedChatComponent.fromText("Example #"+i));
    packets.add(packet);
}

PacketContainer container = new PacketContainer(PacketType.Play.Server.BUNDLE);
container.getPacketBundles().write(0, packets); // <--- This throws exception

// Player player = [...];
ProtocolManager connection = ProtocolLibrary.getProtocolManager();
connection.sendServerPacket(player, container);
```

**Expected behavior**
It should instantiate and send the bundled packet. Instead, it throws an exception:
```
java.lang.IllegalStateException: Unable to set value of field private final java.lang.Iterable net.minecraft.network.protocol.BundlePacket.a
        at com.comphenix.protocol.reflect.accessors.DefaultFieldAccessor.set(DefaultFieldAccessor.java:44) ~[ProtocolLib-5.0.0-SNAPSHOT-b636.jar:?]
        at com.comphenix.protocol.reflect.StructureModifier.writeInternal(StructureModifier.java:359) ~[ProtocolLib-5.0.0-SNAPSHOT-b636.jar:?]
        at com.comphenix.protocol.reflect.StructureModifier.write(StructureModifier.java:322) ~[ProtocolLib-5.0.0-SNAPSHOT-b636.jar:?]
        at io.josemmo.bukkit.plugin.renderer.FakeEntity.tryToSendPackets(FakeEntity.java:122) ~[YamipaPlugin.jar:?]
        at io.josemmo.bukkit.plugin.renderer.FakeImage.nextStep(FakeImage.java:542) ~[YamipaPlugin.jar:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) ~[?:?]
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305) ~[?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
        at java.lang.Thread.run(Thread.java:833) ~[?:?]
Caused by: java.lang.ClassCastException: Cannot cast net.minecraft.network.protocol.BundleDelimiterPacket to net.minecraft.network.protocol.BundlePacket
        at java.lang.Class.cast(Class.java:3889) ~[?:?]
        at com.comphenix.protocol.reflect.accessors.DefaultFieldAccessor.set(DefaultFieldAccessor.java:41) ~[ProtocolLib-5.0.0-SNAPSHOT-b636.jar:?]
        ... 10 more
```

<details>
<summary><strong>Version Info</strong></summary>

```
ProtocolLib Dump
Timestamp: 04/24/23 18:58:39

ProtocolLib Version: ProtocolLib v5.0.0-SNAPSHOT-636
Bukkit Version: 1.19.4-R0.1-SNAPSHOT
Server Version: git-Paper-516 (MC: 1.19.4)
Java Version: 17.0.6

ProtocolLib: com.comphenix.protocol.ProtocolLib@593b8127[
  statistics=com.comphenix.protocol.metrics.Statistics@5a08fad4
  packetTask=8
  tickCounter=3327
  configExpectedMod=1
  updater=com.comphenix.protocol.updater.SpigotUpdater@51cc0be8
  redirectHandler=com.comphenix.protocol.ProtocolLib$2@799decf2
  commandProtocol=com.comphenix.protocol.CommandProtocol@7c2a685f
  commandPacket=com.comphenix.protocol.CommandPacket@6169f354
  commandFilter=com.comphenix.protocol.CommandFilter@504b27da
  packetLogging=com.comphenix.protocol.PacketLogging@4627bb5d
  skipDisable=false
  isEnabled=true
  loader=io.papermc.paper.plugin.manager.DummyBukkitPluginLoader@69e04b56
  server=CraftServer{serverName=Paper,serverVersion=git-Paper-516,minecraftVersion=1.19.4}
  file=plugins\ProtocolLib-5.0.0-SNAPSHOT-b636.jar
  description=org.bukkit.plugin.PluginDescriptionFile@4ec16ca7
  pluginMeta=org.bukkit.plugin.PluginDescriptionFile@4ec16ca7
  dataFolder=plugins\ProtocolLib
  classLoader=PluginClassLoader{plugin=ProtocolLib v5.0.0-SNAPSHOT-636, pluginEnabled=true, url=plugins\ProtocolLib-5.0.0-SNAPSHOT-b636.jar}
  naggable=true
  newConfig=YamlConfiguration[path='', root='YamlConfiguration']
  configFile=plugins\ProtocolLib\config.yml
  logger=com.destroystokyo.paper.utils.PaperPluginLogger@5e168c74
]
Manager: com.comphenix.protocol.injector.PacketFilterManager@76b08745[
  plugin=ProtocolLib v5.0.0-SNAPSHOT-636
  server=CraftServer{serverName=Paper,serverVersion=git-Paper-516,minecraftVersion=1.19.4}
  reporter=com.comphenix.protocol.ProtocolLib$1@33516960
  minecraftVersion=(MC: 1.19.4)
  asyncFilterManager=com.comphenix.protocol.async.AsyncFilterManager@38462038
  pluginVerifier=com.comphenix.protocol.injector.PluginVerifier@52596e68
  inboundListeners=com.comphenix.protocol.injector.SortedPacketListenerList@20a151c9
  outboundListeners=com.comphenix.protocol.injector.SortedPacketListenerList@188336d4
  registeredListeners=[io.josemmo.bukkit.plugin.renderer.ItemService@19214e8]
  packetInjector=com.comphenix.protocol.injector.netty.manager.NetworkManagerPacketInjector@114d821c
  playerInjectionHandler=com.comphenix.protocol.injector.netty.manager.NetworkManagerPlayerInjector@87e8473
  networkManagerInjector=com.comphenix.protocol.injector.netty.manager.NetworkManagerInjector@22fbb3ac
  debug=false
  closed=false
  injected=true
]

Listeners:
io.josemmo.bukkit.plugin.renderer.ItemService@19214e8[
]

Plugins Using ProtocolLib:
YamipaPlugin by []
```
</details>

--- a/src/main/java/com/comphenix/protocol/injector/packet/PacketRegistry.java
+++ b/src/main/java/com/comphenix/protocol/injector/packet/PacketRegistry.java
@@ -363,6 +363,9 @@ public class PacketRegistry {
 		// Try the lookup first (may be null, so check contains)
 		Optional<Class<?>> res = REGISTER.typeToClass.get(type);
 		if (res != null) {
+			if(res.isPresent() && MinecraftReflection.isBundleDelimiter(res.get())) {
+				return MinecraftReflection.getPackedBundlePacketClass();
+			}
 			return res;
 		}
 
@@ -375,6 +378,9 @@ public class PacketRegistry {
 
 		// cache it for next time
 		associate(type, clazz);
+		if(clazz != null && MinecraftReflection.isBundleDelimiter(clazz)) {
+			clazz = MinecraftReflection.getPackedBundlePacketClass().orElseThrow(() -> new IllegalStateException("Packet bundle class not found."));
+		}
 		return Optional.ofNullable(clazz);
 	}
 

--- a/src/test/java/com/comphenix/protocol/PacketTypeTest.java
+++ b/src/test/java/com/comphenix/protocol/PacketTypeTest.java
@@ -22,6 +22,7 @@ import com.comphenix.protocol.PacketType.Protocol;
 import com.comphenix.protocol.PacketType.Sender;
 import com.comphenix.protocol.events.PacketContainer;
 import com.comphenix.protocol.injector.packet.PacketRegistry;
+import com.comphenix.protocol.utility.MinecraftReflection;
 import com.comphenix.protocol.utility.MinecraftReflectionTestUtil;
 import java.lang.reflect.Field;
 import java.util.ArrayList;
@@ -31,6 +32,9 @@ import java.util.Map;
 import java.util.Map.Entry;
 import java.util.Set;
 import java.util.TreeMap;
+
+import com.comphenix.protocol.wrappers.BukkitConverters;
+import com.comphenix.protocol.wrappers.WrappedChatComponent;
 import net.minecraft.network.EnumProtocol;
 import net.minecraft.network.protocol.EnumProtocolDirection;
 import net.minecraft.network.protocol.login.PacketLoginInStart;
@@ -356,5 +360,18 @@ public class PacketTypeTest {
 			}
 		}
 		assertFalse(fail, "Packet type(s) failed to instantiate");
+	}
+
+	@Test
+	public void testPacketBundleWriting() {
+		PacketContainer bundlePacket = new PacketContainer(PacketType.Play.Server.BUNDLE);
+		assertEquals(MinecraftReflection.getPackedBundlePacketClass().orElseThrow(() -> new IllegalStateException("Packet Bundle class is not present")), bundlePacket.getHandle().getClass());
+		List<PacketContainer> bundle = new ArrayList<>();
+
+		PacketContainer chatMessage = new PacketContainer(PacketType.Play.Server.SYSTEM_CHAT);
+		chatMessage.getStrings().write(0, WrappedChatComponent.fromText("Test").getJson());
+		chatMessage.getBooleans().write(0, false);
+		bundle.add(chatMessage);
+		bundlePacket.getPacketBundles().write(0, bundle);
 	}
 }

